change the cluster properties on the sources option

IhsenBen opened the pull request:
## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - I Confirm that **my changes do not include backports from Mapbox projects** (unless with compliant license) 
 - fix issue #1384 and #429 
 -  enable users to change cluster properties on the source by adding setClusterOptions in [src/source/geojson-source.ts](https://github.com/maplibre/maplibre-gl-js/blob/v2.4.0/src/source/geojson_source.ts#L68-L346)
 - `setClusterOptions(cluster: boolean, clusterRadius?: number; clusterMaxZoom?: number})`  The value of the clusters properties.
 
Exemple:
```js
map.addSource('earthquakes', {
                type: 'geojson',
                data:
                    'https://maplibre.org/maplibre-gl-js-docs/assets/earthquakes.geojson',
                cluster: true,
            });
///////////////
         map.getSource('earthquakes').setClusterOptions(cluster:false);
        map.getSource('earthquakes').setClusterOptions(clust: true, clusterRadius: 80, clusterMaxZoom: 50 });
/////////////
```

HarelM commented:
Thanks for taking the time to contribute!!
Unfortunately, I don't think this code works.
Geojson source uses supercluster and a worker to split the data into tiles.
I haven't seen any changes in that code, which I would expect.
Did you test it?
Please also add the relevant tests fort this use case.

IhsenBen commented:
> Thanks for taking the time to contribute!!
> 
> Unfortunately, I don't think this code works.
> 
> Geojson source uses supercluster and a worker to split the data into tiles.
> 
> I haven't seen any changes in that code, which I would expect.
> 
> Did you test it?
> 
> Please also add the relevant tests fort this use case.

Hey, Sorry to hear that, I’m only changing the property ( not removing the clusters ), I tested it with the CSS/Js build and it was working.
Besides testing, Can you please tell me more about what you would expect from this change? 
Thx in advance

HarelM commented:
Can you share the code that you used for testing this?
Since the addition to the code only changes the `_options` and since changing these don't call the `_updateWorkerData` and doesn't change `this.workerOptions` I don't see how this would work. I might be wrong though :-)
The only thing I can think of, is if you set this before the source is added to the map and so this option is used only on startup once.
Can you share a working example, or maybe a movie where you can toggle on and off the cluster?

IhsenBen commented:
Oh I see, that does make sense, because I had this use case while I was reloading it with React, I tested the build with a simple HTML code with buttons like this (so as you said this only runs once):
```HTML
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="maplibre-gl.js"></script>
    <link rel="stylesheet" href="maplibre-gl.css">
    <title>Document</title>
</head>

<body>
    <style>
        body {
            margin: 0;
            padding: 0;
        }

        .container {
            position: relative;
            width: 100%;
            height: 100vh;
            display: flex;
            flex-direction: column;
            justify-content: center;
        }

        #map {
            position: absolute;
            top: 0;
            bottom: 0;
            width: 100%;
        }
        .buttons{
            position: absolute;
            top: 0;
            left: 0;
            z-index: 1;
        }
    </style>
    <h1>Webpack</h1>
    <div class="container">
        <div class="buttons">
            <button id="on">Turn On</button>
            <button id="off">Turn Off</button>
        </div>
        <div class="mapcontainer">
            <div id="map"></div>
        </div>

    </div>
    <script>

        var map = new maplibregl.Map({
            container: 'map',
            style:
                'https://api.maptiler.com/maps/streets/style.json?key=get_your_own_OpIi9ZULNHzrESv6T2vL',
            center: [-103.59179687498357, 40.66995747013945],
            zoom: 3
        });

        map.on('load', function () {
            // Add a new source from our GeoJSON data and
            // set the 'cluster' option to true. GL-JS will
            // add the point_count property to your source data.
            map.addSource('earthquakes', {
                type: 'geojson',
                // Point to GeoJSON data. This example visualizes all M1.0+ earthquakes
                // from 12/22/15 to 1/21/16 as logged by USGS' Earthquake hazards program.
                data:
                    'https://maplibre.org/maplibre-gl-js-docs/assets/earthquakes.geojson',
                cluster: true,
                clusterMaxZoom: 50, // Max zoom to cluster points on
                clusterRadius: 50 // Radius of each cluster when clustering points (defaults to 50)
            });

            map.addLayer({
                id: 'clusters',
                type: 'circle',
                source: 'earthquakes',
                filter: ['has', 'point_count'],
                paint: {
                    // Use step expressions (https://maplibre.org/maplibre-gl-js-docs/style-spec/#expressions-step)
                    // with three steps to implement three types of circles:
                    //   * Blue, 20px circles when point count is less than 100
                    //   * Yellow, 30px circles when point count is between 100 and 750
                    //   * Pink, 40px circles when point count is greater than or equal to 750
                    'circle-color': [
                        'step',
                        ['get', 'point_count'],
                        '#51bbd6',
                        100,
                        '#f1f075',
                        750,
                        '#f28cb1'
                    ],
                    'circle-radius': [
                        'step',
                        ['get', 'point_count'],
                        20,
                        100,
                        30,
                        750,
                        40
                    ]
                }
            });

            map.addLayer({
                id: 'cluster-count',
                type: 'symbol',
                source: 'earthquakes',
                filter: ['has', 'point_count'],
                layout: {
                    'text-field': '{point_count_abbreviated}',
                    'text-font': ['DIN Offc Pro Medium', 'Arial Unicode MS Bold'],
                    'text-size': 12
                }
            });

            map.addLayer({
                id: 'unclustered-point',
                type: 'circle',
                source: 'earthquakes',
                filter: ['!', ['has', 'point_count']],
                paint: {
                    'circle-color': '#11b4da',
                    'circle-radius': 4,
                    'circle-stroke-width': 1,
                    'circle-stroke-color': '#fff'
                }
            });

            // inspect a cluster on click
            map.on('click', 'clusters', function (e) {
                var features = map.queryRenderedFeatures(e.point, {
                    layers: ['clusters']
                });
                var clusterId = features[0].properties.cluster_id;
                map.getSource('earthquakes').getClusterExpansionZoom(
                    clusterId,
                    function (err, zoom) {
                        if (err) return;

                        map.easeTo({
                            center: features[0].geometry.coordinates,
                            zoom: zoom
                        });
                    }
                );
            });

            // When a click event occurs on a feature in
            // the unclustered-point layer, open a popup at
            // the location of the feature, with
            // description HTML from its properties.
            map.on('click', 'unclustered-point', function (e) {
                var coordinates = e.features[0].geometry.coordinates.slice();
                var mag = e.features[0].properties.mag;
                var tsunami;

                if (e.features[0].properties.tsunami === 1) {
                    tsunami = 'yes';
                } else {
                    tsunami = 'no';
                }

                // Ensure that if the map is zoomed out such that
                // multiple copies of the feature are visible, the
                // popup appears over the copy being pointed to.
                while (Math.abs(e.lngLat.lng - coordinates[0]) > 180) {
                    coordinates[0] += e.lngLat.lng > coordinates[0] ? 360 : -360;
                }

                new maplibregl.Popup()
                    .setLngLat(coordinates)
                    .setHTML(
                        'magnitude: ' + mag + '<br>Was there a tsunami?: ' + tsunami
                    )
                    .addTo(map);

            });


            const turnOnBtn = document.getElementById('on');
            const turnOffBtn = document.getElementById('off');

            turnOnBtn.addEventListener('click', () => {
                map.setLayoutProperty('clusters', 'visibility', 'visible');
                map.getSource('earthquakes').setCluster(true);
                console.log(map.getSource('earthquakes')._options.cluster)
            });

            turnOffBtn.addEventListener('click', () => {
                map.setLayoutProperty('clusters', 'visibility', 'none');
                map.getSource('earthquakes').setCluster(false);
                console.log(map.getSource('earthquakes')._options.cluster)
            });

        });

    </script>

</body>

</html>
```

IhsenBen commented:
I updated the pull request and added more options, I'm open to your suggestions or requirements 😃

HarelM commented:
Thanks for the update, please see if you can add tests to this scenario.
I would also recommend to reduce the changes in the changelog, did you format it?

IhsenBen commented:
Hi,
Sorry about the auto formatting I had it "on save", So :
- I added a unit test .
- cleaned my log on changelog .
- and squashed my commits.
I hope it's fine now.
Also please don't hesitate to ask if you want more changes.
thx in advance

HarelM commented:
This looks great!
Can you please add types to `_options` and `workerOptions`?
It's unclear what can be set there, what is passed to the worker, what is kept in the source etc.
I'm guessing some types can be fetched from supercluster library, others might be needed to be added manually.
Fixing this will allow us to see if there are places where the current change is missing (i.e. more places where we need to change the values of the saved members).
Otherwise, this looks good.

IhsenBen commented:
> This looks great! Can you please add types to `_options` and `workerOptions`? 
I don't understand, I already typed the arguments of setCluster().
can you please tell me what do you mean exactly?

HarelM commented:
The class members have a type `any`. Since you changes the code that uses then I was hoping you could add types to them to improve readability and future maintenance.

IhsenBen commented:
> The class members have a type `any`. Since you changes the code that uses them, I was hoping you could add types to them to improve readability and future maintenance.


I rebased my branch and added the type for _options with optional properties for flexibility, also tbh I didn't want to bother with workerOptions because this has a lot values and I didn't want to mess anything up.

Hope this will do it :D 
thx for your time and consideration

HarelM commented on src/source/geojson_source.ts:
```diff
@@ -18,6 +18,14 @@ export type GeoJSONSourceOptions = GeoJSONSourceSpecification & {
     workerOptions?: any;
     collectResourceTiming: boolean;
 }
+export type _options = {
```
This should get a proper name like `GeoJsonSourceOptions`

HarelM commented on src/source/geojson_source.ts:
```diff
@@ -196,6 +204,28 @@ class GeoJSONSource extends Evented implements Source {
+    /**
+     * To disable/enable clustering on the source options
+     * @param {boolean} cluster Whether to enable clustering.
+     * @param {Object} options   Options object
```
Better to have this as a type and reference this type in the jsdocs comment.

HarelM commented on src/source/geojson_source.ts:
```diff
@@ -196,6 +204,28 @@ class GeoJSONSource extends Evented implements Source {
+     * map.getSource('some id').setCluster(true);
+     *
+     */
+    setCluster(cluster: boolean, options?: {clusterRadius?: number; clusterMaxZoom?: number}) {
```
I would consider calling this `setClusterOptions` to align with what this method does.

HarelM commented:
Would be nice to type the worker option as best as you could at this point in time, we can always improve this in the future. Anything is better than `any`.
If this is too much to ask then please open an issue on it so someone else could pick this up.
Boy Scout role etc... 😀

IhsenBen commented:
> Would be nice to type the worker option as best as you could at this point in time, we can always improve this in the future. Anything is better than `any`. If this is too much to ask then please open an issue on it so someone else could pick this up. Boy Scout role etc... 😀

Hey
I updated my work and added the extra types. ( I can move the types to a different file if you suggest one)
made them flexible with optional properties, and so far seems to be passing all the tests. 
hopefully, this will do it.
looking forward for your feedback

HarelM commented on src/source/geojson_source.ts:
```diff
@@ -19,6 +19,45 @@ export type GeoJSONSourceOptions = GeoJSONSourceSpecification & {
+    collectResourceTiming?: boolean;
+}
+
+export type setClusterOptions = {
```
Much better! Can you please use capital letters at the begging of the type (PascalCasing)?

IhsenBen replied:
Ah my bad.
Here you go, just pushed these fixes :D
Good Weekend 👍

IhsenBen commented:
Thanks Harel, this was my first valuable open-source contribution.
good team work

HarelM commented:
First of many hopefully 😀
